### PR TITLE
[tool] blockchain  time machine

### PR DESCRIPTION
**Pull request by huof6829**
# Description
timemachine is a command-line tool to go back specific block height. user can try the specific-height data is right or not.

steps:
1. prepare yaml: 
     refer to iotex-bootstrap to download config.yaml and genesis.yaml to ./bin/timemachine/etc/
2. compile: 
     make build-timemachine && cd ./bin/timemachine
3. download datas to ./bin/timemachine/data/:  
    ./timemachine download [mainnet/testnet] [height]           
4. edit ./etc/config.yaml and set path of chainDB/trieDB/pollDB after download successful
5. query current height of chian and trie
    ./bin/timemachine get --config-path=./etc/config.yaml --genesis-path=./etc/genesis.yaml                                 
6. try a specific height,  validate the height without committed  
     ./bin/timemachine try [height] --config-path=./etc/config.yaml  --genesis-path=./etc/genesis.yaml                   

Fixes https://github.com/iotexproject/iotex-core/issues/3677

## Type of change
Please delete options that are not relevant.
- [] Bug fix (non-breaking change which fixes an issue)
- [] New feature (non-breaking change which adds functionality)
- [] Code refactor or improvement
- [] Breaking change (fix or feature that would cause a new or changed behavior of existing functionality)
- [] This change requires a documentation update

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration
- [] make test
- [] fullsync
- [] Other test (please specify)

**Test Configuration**:
- Firmware version:
- Hardware:
- Toolchain:
- SDK:

# Checklist:
- [] My code follows the style guidelines of this project
- [] I have performed a self-review of my code
- [] I have commented my code, particularly in hard-to-understand areas
- [] I have made corresponding changes to the documentation
- [] My changes generate no new warnings
- [] I have added tests that prove my fix is effective or that my feature works
- [] New and existing unit tests pass locally with my changes
- [] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: tools/timemachine/config.yaml [hunk 1]
```diff
@@ -0,0 +1,18 @@
+chain:
+  chainDBPath: /Users/u/Downloads/fullsync_mainnet_14m25_chain-00000013.db
+  trieDBPath: /Users/u/Downloads/fullsync_mainnet_14m25_trie.db
```
Comment (dustinxie) [inline tools/timemachine/config.yaml:3]: cannot use fixed path/filename

File: tools/timemachine/internal/cmd/downloadHeight.go [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+		cmd.SilenceUsage = true
+		height, err := strconv.ParseUint(args[0], 10, 64)
+		if err != nil {
+			return errors.Wrapf(err, "Failed to convert input height: %s.", args[0])
+		}
+		if height == 0 {
+			return errors.New("input height cannot be 0.")
+		}
+		var (
+			heighM    = float64(height) / float64(1024) / float64(1024)
```
Comment (dustinxie) [inline tools/timemachine/internal/cmd/downloadHeight.go:47]: height is calculated by 1million = 1000*1000

File: tools/timemachine/internal/cmd/downloadHeight.go [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+			heightDir string
+			destDir   = args[1]
```
Comment (dustinxie) [inline tools/timemachine/internal/cmd/downloadHeight.go:49]: destDir should not be an arg, it should be determined by the height

File: tools/timemachine/internal/cmd/downloadHeight.go [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+			bucket    = "blockchain-golden"
+			prefix    = "fullsync/mainnet/"
```
Comment (dustinxie) [inline tools/timemachine/internal/cmd/downloadHeight.go:51]: 1. bucket and prefix can be a local constant
2. one var "blockchain-golden/fullsync/mainnet/" should be good enough

File: tools/timemachine/internal/cmd/downloadHeight.go [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+						if v > curr {
+							curr = v
+						}
+					}
+				}
+			}
+		}
+		latest := curr + 0.25
+
+		if heighM < 8 {
```
Comment (dustinxie) [inline tools/timemachine/internal/cmd/downloadHeight.go:87]: no need to divide by 100000, just use the height
0-8000000
8000000-12000000
12000000-13000000
13000000-13250000
etc.

File: Makefile [hunk 6]
```diff
@@ -101,6 +102,10 @@ build-staterecoverer:
 build-readtip:
 	$(GOBUILD) -o ./bin/$(BUILD_TARGET_READTIP) -v ./tools/readtip
 
+.PHONY: build-timemachine
+build-timemachine:
```
Comment (dustinxie) [inline Makefile:106]: do not need this line?

File: tools/timemachine/internal/cmd/downloadHeight.go [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+	"cloud.google.com/go/storage"
+	"github.com/pkg/errors"
+	"github.com/spf13/cobra"
+	"google.golang.org/api/iterator"
+)
+
+// const value
+var (
+	GcpTimeout  = time.Second * 60
+	ErrNotExist = errors.New("height is not existed.")
```
Comment (dustinxie) [inline tools/timemachine/internal/cmd/downloadHeight.go:29]: "height does not exist"

File: tools/timemachine/internal/cmd/downloadHeight.go [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+						if v > curr {
+							curr = v
+						}
+					}
+				}
+			}
+		}
+		latest := curr + 250000
+
+		if height < 8000000 {
```
Comment (dustinxie) [inline tools/timemachine/internal/cmd/downloadHeight.go:87]: move this code block to `func heightToDownloadDir()`

File: tools/timemachine/internal/cmd/downloadHeight.go [hunk 9]
```diff
@@ -0,0 +1,228 @@
...
+			heightDir = "0m"
+		} else if height >= 8000000 && height < 12000000 {
+			heightDir = "8m"
+		} else if height >= 12000000 && height < 13000000 {
+			heightDir = "12m"
+		} else if height >= 13000000 && height < latest {
+			inter := height / 1000000
+			deci := height - inter
```
Comment (dustinxie) [inline tools/timemachine/internal/cmd/downloadHeight.go:95]: i think should be `deci := height - inter * 1000000`?

File: tools/timemachine/internal/cmd/downloadHeight.go [hunk 10]
```diff
@@ -0,0 +1,228 @@
...
+	if err != nil {
+		return errors.Wrap(err, "Failed to storage.NewClient.")
+	}
+	defer client.Close()
+	client.SetRetry(storage.WithPolicy(storage.RetryAlways))
+
+	ctx, cancel := context.WithTimeout(ctx, GcpTimeout)
+	defer cancel()
+
+	if err = mkdirIfNotExist(filepath.Dir(destFileName)); err != nil {
```
Comment (dustinxie) [inline tools/timemachine/internal/cmd/downloadHeight.go:148]: if destFileName does not exist, it is an error?

File: tools/timemachine/internal/cmd/syncHeight.go [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+			protocol.BlockCtx{
+				BlockHeight:    blk.Height(),
+				BlockTimeStamp: blk.Timestamp(),
+				Producer:       producer,
+				GasLimit:       gCtx.BlockGasLimit,
+			},
+		), blk); err == nil {
+			break
+		}
+		if errors.Cause(err) == block.ErrDeltaStateMismatch {
```
Comment (dustinxie) [inline tools/timemachine/internal/cmd/syncHeight.go:119]: this error can only be ignored before Greenland (or Iceland) height

File: tools/timemachine/config.yaml (hunk 1 above)
Comment (huof6829) [reply tools/timemachine/config.yaml]: fixed

File: tools/timemachine/internal/cmd/downloadHeight.go (hunk 2 above)
Comment (huof6829) [reply tools/timemachine/internal/cmd/downloadHeight.go]: fixed

File: tools/timemachine/internal/cmd/downloadHeight.go (hunk 5 above)
Comment (huof6829) [reply tools/timemachine/internal/cmd/downloadHeight.go]: fixed

File: tools/timemachine/internal/cmd/downloadHeight.go (hunk 7 above)
Comment (huof6829) [reply tools/timemachine/internal/cmd/downloadHeight.go]: fixed

File: tools/timemachine/internal/cmd/downloadHeight.go (hunk 8 above)
Comment (huof6829) [reply tools/timemachine/internal/cmd/downloadHeight.go]: fixed

File: tools/timemachine/internal/cmd/downloadHeight.go (hunk 9 above)
Comment (huof6829) [reply tools/timemachine/internal/cmd/downloadHeight.go]: right.

File: tools/timemachine/internal/cmd/syncHeight.go (hunk 11 above)
Comment (huof6829) [reply tools/timemachine/internal/cmd/syncHeight.go]: ok

File: tools/timemachine/internal/cmd/downloadHeight.go (hunk 10 above)
Comment (huof6829) [reply tools/timemachine/internal/cmd/downloadHeight.go]: if not exist, try to mkdir,  if error, return error.

File: Makefile (hunk 6 above)
Comment (huof6829) [reply Makefile:106]: add better. I add it  referring to other commands.

File: tools/timemachine/internal/cmd/downloadHeight.go (hunk 4 above)
Comment (huof6829) [reply tools/timemachine/internal/cmd/downloadHeight.go]: fixed

File: action/protocol/context.go [hunk 12]
```diff
@@ -76,6 +79,16 @@ type (
 		Nonce uint64
 	}
 
+	// TestCtx is ctx used in test or tool
+	TestCtx struct {
+		// DisableCheckIndexer disables to check indexer
+		DisableCheckIndexer bool
```
Comment (dustinxie) [inline action/protocol/context.go:85]: don't need this flag?
it's always true for both the `play` and `commit` command

Comment (dustinxie) [reply action/protocol/context.go:85]: in other words, the existence of `TestCtx` indicates user wants to disable check indexer

File: tools/timemachine/internal/cmd/play.go [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+			return err
+		}
+		svr, err := newMiniServer(miniServerConfig(), WithStopHeight(stopHeight))
+		if err != nil {
+			return err
+		}
+		if err = svr.checkIndexer(); err != nil {
+			return err
+		}
+		log.L().Info("play the height's block successfully.")
```
Comment (dustinxie) [inline tools/timemachine/internal/cmd/play.go:35]: "successfully played block %d", stopHeight

File: tools/timemachine/internal/cmd/commit.go [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+			return err
+		}
+		svr, err := newMiniServer(miniServerConfig(), WithStopHeight(stopHeight), EnableCommitBlock())
+		if err != nil {
+			return err
+		}
+		if err = svr.checkIndexer(); err != nil {
+			return err
+		}
+		log.L().Info("commit the height's block successfully.")
```
Comment (dustinxie) [inline tools/timemachine/internal/cmd/commit.go:35]: "successfully committed block %d", stopHeight

File: state/factory/statedb.go [hunk 15]
```diff
@@ -353,6 +353,10 @@ func (sdb *stateDB) PutBlock(ctx context.Context, blk *block.Block) error {
 		)
 	}
 
+	if testCfg, ok := protocol.GetTestCtx(ctx); ok &&
+		blk.Height() == testCfg.StopHeight && !testCfg.CommitBlock {
+		return nil
+	}
```
Comment (dustinxie) [inline state/factory/statedb.go:359]: factory.go

File: action/protocol/context.go (hunk 12 above)
Comment (huof6829) [reply action/protocol/context.go]: fixed

File: tools/timemachine/internal/cmd/commit.go (hunk 14 above)
Comment (huof6829) [reply tools/timemachine/internal/cmd/commit.go]: fixed

File: state/factory/statedb.go (hunk 15 above)
Comment (huof6829) [reply state/factory/statedb.go:359]: added

File: tools/timemachine/internal/cmd/download.go [hunk 16]
```diff
@@ -0,0 +1,196 @@
...
+
+		heightDir := genHeightDir(height)
+		log.S().Infof("download the height's dir: %s", heightDir)
+
+		objs, err := listFiles(_bucket, _objPrefix+heightDir+"/", "/")
+		if err != nil {
+			return err
+		}
+		if len(objs) == 0 {
+			objs, err = listFiles(_bucket, _objPrefix+"latest"+"/", "/")
```
Comment (dustinxie) [inline tools/timemachine/internal/cmd/download.go:65]: return `latest` in `genHeightDir`, so don't need to check here

File: tools/timemachine/internal/cmd/download.go [hunk 17]
```diff
@@ -0,0 +1,196 @@
...
+		} else {
+			heightDir = interStr + "75"
+		}
+	}
+	return
+}
+
+func mkdirIfNotExist(destDir string) error {
+	if _, err := os.Stat(destDir); os.IsNotExist(err) {
+		if err = os.MkdirAll(destDir, 0744); err != nil {
```
Comment (dustinxie) [inline tools/timemachine/internal/cmd/download.go:191]: we should not create dir? if the path does not exist, something is wrong?

Comment (huof6829) [reply tools/timemachine/internal/cmd/download.go:191]: mkdir `fullsync/mainnet/13m/`   to  keep with dir in gcp

Comment (huof6829) [reply tools/timemachine/internal/cmd/download.go:191]: mkdir `fullsync/mainnet/13m/`   to  keep with dir in gcp

File: tools/timemachine/internal/cmd/download.go (hunk 16 above)
Comment (huof6829) [reply tools/timemachine/internal/cmd/download.go]: fixed

File: tools/timemachine/internal/cmd/download.go (hunk 17 above)
Comment (dustinxie) [reply tools/timemachine/internal/cmd/download.go:187]: timemachine/13m

File: chainservice/builder.go [hunk 18]
```diff
@@ -147,10 +171,6 @@ func (builder *Builder) createFactory(forTest bool) (factory.Factory, error) {
 		if forTest {
 			return factory.NewStateDB(factoryCfg, db.NewMemKVStore(), factory.RegistryStateDBOption(builder.cs.registry))
 		}
-		opts := []factory.StateDBOption{
-			factory.RegistryStateDBOption(builder.cs.registry),
-			factory.DefaultPatchOption(),
```
Comment (dustinxie) [inline chainservice/builder.go:152]: `factory.RegistryStateDBOption(builder.cs.registry)` is common option, so keep it

File: chainservice/builder.go [hunk 19]
```diff
@@ -159,7 +179,20 @@ func (builder *Builder) createFactory(forTest bool) (factory.Factory, error) {
...
+			return minifactory.NewStateDB(factoryCfg, dao, minifactory.RegistryStateDBOption(builder.cs.registry), minifactory.WithStopHeightStateDBOption(builder.stopHeight))
+		case minifactory.Commit:
+			return minifactory.NewStateDB(factoryCfg, dao, minifactory.RegistryStateDBOption(builder.cs.registry), minifactory.CommitBlockStateDBOption())
+		case minifactory.Get:
+			return minifactory.NewStateDB(factoryCfg, dao, minifactory.RegistryStateDBOption(builder.cs.registry))
+		default:
+			opts := []factory.StateDBOption{
+				factory.RegistryStateDBOption(builder.cs.registry),
+				factory.DefaultPatchOption(),
+			}
```
Comment (dustinxie) [inline chainservice/builder.go:193]:
```
case mini.Try:
    opts = append(opts, WithStopHeightStateDBOption)
case mini.Commit:
    opts = append(opts, CommitBlockStateDBOption)
case mini.Get:
default:
    opts = append(opts, DefaultPatchOption)
```

File: tools/timemachine/minidao/blockdao.go [hunk 20]
```diff
@@ -0,0 +1,142 @@
...
+	if err != nil {
+		return err
+	}
+	atomic.StoreUint64(&dao.tipHeight, tipHeight)
+	return nil
+}
+
+func (dao *blockDAO) Stop(ctx context.Context) error { return dao.lifecycle.OnStop(ctx) }
+
+func (dao *blockDAO) Height() (uint64, error) {
```
Comment (dustinxie) [inline tools/timemachine/minidao/blockdao.go:80]: these funcs can be omitted if you using embedded `blockDao` struct?

File: tools/timemachine/minidao/blockdao.go [hunk 21]
```diff
@@ -0,0 +1,142 @@
...
+	"github.com/iotexproject/iotex-core/pkg/lifecycle"
+	"github.com/iotexproject/iotex-core/pkg/log"
+)
+
+type (
+	blockDAO struct {
+		blockStore filedao.FileDAO
+		indexers   []blockdao.BlockIndexer
+		lifecycle  lifecycle.Lifecycle
+		tipHeight  uint64
```
Comment (dustinxie) [inline tools/timemachine/minidao/blockdao.go:32]: struct {
*blockdao.BlockDao
}

File: tools/timemachine/internal/cmd/download.go (hunk 17 above)
Comment (huof6829) [reply tools/timemachine/internal/cmd/download.go:184]: fixed

File: tools/timemachine/etc/config.yaml [hunk 22]
```diff
@@ -0,0 +1,66 @@
+network:
```
Comment (CoderZhi) [inline tools/timemachine/etc/config.yaml:1]: simplify this file

File: tools/timemachine/etc/genesis.yaml [hunk 23]
```diff
@@ -0,0 +1,98 @@
+account:
```
Comment (CoderZhi) [inline tools/timemachine/etc/genesis.yaml:1]: delete this file

File: tools/timemachine/minifactory/statedb.go [hunk 24]
```diff
@@ -0,0 +1,78 @@
+// Copyright (c) 2022 IoTeX Foundation
```
Comment (CoderZhi) [inline tools/timemachine/minifactory/statedb.go:1]: what's the problem of using the original statedb implementation with options?

Comment (huof6829) [reply tools/timemachine/minifactory/statedb.go:1]: if use options, statedb struct will add member variable.  But the member variable is just used in tools, no matter with blockchain. So  inherit from statedb, just make changes in tools/timemachine.
Or  using ctx can be better?  like 
```
func (sdb *statedb) PutBlock() {
         ...
	if testCfg, ok := protocol.GetTestCtx(ctx); ok &&
		blk.Height() == testCfg.StopHeight && !testCfg.CommitBlock {
		return nil
	}
        ws.Commit(ctx)
        ...
}
```  
see the former commit:   [delete testCtx and realize in tools](https://github.com/iotexproject/iotex-core/pull/3692/commits/6756dc67d2a169d6515ffc08fc93cff3a5f8ce64)

Comment (dustinxie) [reply tools/timemachine/minifactory/statedb.go:1]: i think just use options, make a clear comment this flag/var is only used by timemachine tools

Comment (huof6829) [reply tools/timemachine/minifactory/statedb.go]: use option to do and fixed

File: state/factory/factory.go [hunk 25]
```diff
@@ -471,6 +489,9 @@ func (sf *factory) PutBlock(ctx context.Context, blk *block.Block) error {
 		)
 	}
 
+	if blk.Height() == sf.stopHeight && !sf.commitBlock {
```
Comment (CoderZhi) [inline state/factory/factory.go:492]: if `blk.Height() == sf.stopHeight` and `sf.commitBlock == true`, block will be committed, and move to next height.
If we want to stop at height `h`, and then commit block `h`, we can easily update `stopHeight` to `h+1`

File: chainservice/builder.go [hunk 26]
```diff
@@ -33,23 +38,43 @@ import (
...
-// Builder is a builder to build chainservice
-type Builder struct {
-	cfg config.Config
-	cs  *ChainService
+type (
+	// Builder is a builder to build chainservice
+	Builder struct {
+		cfg           config.Config
+		cs            *ChainService
+		opTimeMachine int
```
Comment (CoderZhi) [inline chainservice/builder.go:49]: not needed

File: blockchain/blockdao/blockdao.go [hunk 27]
```diff
@@ -96,10 +100,13 @@ func (dao *BlockDAODB) StartDB(ctx context.Context) error {
 		return err
 	}
 	atomic.StoreUint64(&dao.tipHeight, tipHeight)
-	return nil
+	if dao.timemachine {
```
[the remark this answers is not in the document]
Comment (huof6829) [reply blockchain/blockdao/blockdao.go:103]: `get` command should not do `checkIndexers`,  `try` and `commit` command need. so add this check

File: state/factory/factory.go (hunk 25 above)
Comment (huof6829) [reply state/factory/factory.go]: I see. `commit`  command is useless, and delete it

File: blockchain/blockdao/blockdao.go (hunk 27 above)
[the remark this answers is not in the document]
Comment (huof6829) [reply blockchain/blockdao/blockdao.go]: fixed

File: blockchain/blockdao/blockdao.go [hunk 28]
```diff
@@ -55,17 +55,28 @@ type (
 		bodyCache    cache.LRUCache
 		footerCache  cache.LRUCache
 		tipHeight    uint64
+		timemachine  bool
```
[the remark this answers is not in the document]
Comment (huof6829) [reply blockchain/blockdao/blockdao.go]: fixed

File: chainservice/builder.go (hunk 26 above)
Comment (huof6829) [reply chainservice/builder.go]: fixed

File: chainservice/builder.go [hunk 29]
```diff
@@ -33,23 +38,34 @@ import (
...
+// WithStopHeightBuilderOption sets the stopHeight
+func WithStopHeightBuilderOption(stopHeight uint64) BuilderOption {
+	return func(svr *Builder) {
+		svr.stopHeight = stopHeight
+	}
 }
 
 // NewBuilder creates a new chainservice builder
-func NewBuilder(cfg config.Config) *Builder {
+func NewBuilder(cfg config.Config, opts ...BuilderOption) *Builder {
```
Comment (CoderZhi) [inline chainservice/builder.go:63]: `opts` is better to be used in function `Build`.

File: tools/timemachine/internal/cmd/download.go [hunk 30]
```diff
@@ -0,0 +1,189 @@
...
+		} else if deci >= 250000 && deci < 500000 {
+			heightDir = interStr + "25"
+		} else if deci >= 500000 && deci < 750000 {
+			heightDir = interStr + "50"
+		} else {
+			heightDir = interStr + "75"
+		}
+	} else {
+		heightDir = "latest"
+	}
```
[the remark this answers is not in the document]
Comment (CoderZhi) [reply tools/timemachine/internal/cmd/download.go:177]: pls address this comment

File: blockchain/blockdao/blockdao.go [hunk 31]
```diff
@@ -55,17 +55,28 @@ type (
 		bodyCache    cache.LRUCache
 		footerCache  cache.LRUCache
 		tipHeight    uint64
+		stopHeight   uint64 // for tools/timemachine
```
Comment (CoderZhi) [inline blockchain/blockdao/blockdao.go:58]: why we need a stop height for blockdao?

Comment (huof6829) [reply blockchain/blockdao/blockdao.go:58]: `if err := checker.CheckIndexer(ctx, indexer, 0, func(height uint64) {`   
0 means get  `tipHeight`

File: tools/timemachine/internal/cmd/download.go (hunk 30 above)
[the remark this answers is not in the document]
Comment (huof6829) [reply tools/timemachine/internal/cmd/download.go:177]: download folder such as `13m` is enough. folder contains db files.  the folder name is named by dividing height by 25000.  the naming rule is fixed.

File: chainservice/builder.go (hunk 29 above)
Comment (huof6829) [reply chainservice/builder.go]: ok. fixed

File: tools/timemachine/internal/cmd/download.go (hunk 30 above)
[the remark this answers is not in the document]
Comment (huof6829) [reply tools/timemachine/internal/cmd/download.go]: I see. download all file names and parse into a slice and find the nearest.

File: chainservice/builder.go [hunk 32]
```diff
@@ -115,8 +129,11 @@ func (builder *Builder) BuildForSubChain() (*ChainService, error) {
 }
 
 // Build builds a chainservice
-func (builder *Builder) Build() (*ChainService, error) {
+func (builder *Builder) Build(opts ...BuilderOption) (*ChainService, error) {
 	builder.createInstance()
+	for _, opt := range opts {
+		opt(builder)
```
Comment (CoderZhi) [inline chainservice/builder.go:135]: stop height could be just an opt value, which doesn't need to be a variable of `Builder`.

File: blockchain/blockdao/blockdao.go (hunk 31 above)
Comment (CoderZhi) [reply blockchain/blockdao/blockdao.go:58]: I still do not get it.

File: chainservice/builder.go (hunk 32 above)
Comment (huof6829) [reply chainservice/builder.go]: fixed

File: blockchain/blockdao/blockdao.go (hunk 31 above)
Comment (huof6829) [reply blockchain/blockdao/blockdao.go]: fixed.  delete stop height in blockDAO which is repeated and meaningless.

File: chainservice/builder.go [hunk 33]
```diff
@@ -116,9 +117,9 @@ func (builder *Builder) BuildForSubChain() (*ChainService, error) {
 }
 
 // Build builds a chainservice
-func (builder *Builder) Build() (*ChainService, error) {
+func (builder *Builder) Build(opts ...factory.StateDBOption) (*ChainService, error) {
```
Comment (CoderZhi) [inline chainservice/builder.go:120]: Use `BuilderOption` instead

File: tools/timemachine/internal/cmd/root.go [hunk 34]
```diff
@@ -0,0 +1,40 @@
...
+)
+
+// rootCmd represents the base command when called without any subcommands
+var rootCmd = &cobra.Command{
+	Use: "timemachine [command] [flags]",
+	Short: `timemachine is a command-line interface to go back specific block height.
+steps:
+1. mkdir ./tools/timemachine/etc
+   curl https://raw.githubusercontent.com/iotexproject/iotex-bootstrap/v1.8.4/config_mainnet.yaml > ./tools/timemachine/etc/config.yaml
+   curl https://raw.githubusercontent.com/iotexproject/iotex-bootstrap/v1.8.4/genesis_mainnet.yaml > ./tools/timemachine/etc/genesis.yaml
```
Comment (CoderZhi) [inline tools/timemachine/internal/cmd/root.go:21]: `2. refer to iotex-bootstrap to download config and genesis`
else we need to update it for every release

File: tools/timemachine/internal/cmd/get.go [hunk 35]
```diff
@@ -0,0 +1,65 @@
...
+
+// get represents the get command
+var get = &cobra.Command{
+	Use:   "get",
+	Short: "Show current height of trie.db and chain.db",
+	Args:  cobra.NoArgs,
+	RunE: func(cmd *cobra.Command, args []string) error {
+		cmd.SilenceUsage = true
+
+		cfg := miniserver.Config()
```
Comment (CoderZhi) [inline tools/timemachine/internal/cmd/get.go:30]: config file path should be from command parameter, same for genesis file path

File: tools/timemachine/internal/cmd/download.go [hunk 36]
```diff
@@ -0,0 +1,219 @@
...
+		wg.Wait()
+
+		log.S().Infof("successful download height %d", height)
+		return nil
+	},
+}
+
+// downloadFile downloads an object to a file.
+func downloadFile(bucket, object, dbpath string) error {
+	dbpath = fmt.Sprintf("./tools/timemachine/data/%s", dbpath)
```
Comment (CoderZhi) [inline tools/timemachine/internal/cmd/download.go:118]: don't hard code file path

File: tools/timemachine/internal/cmd/download.go [hunk 37]
```diff
@@ -0,0 +1,219 @@
...
+	"google.golang.org/api/iterator"
+
+	"github.com/iotexproject/iotex-core/pkg/log"
+)
+
+// const
+const (
+	_gcpTimeout = time.Second * 60
+	_bucket     = "blockchain-golden"
+	_objPrefix  = "fullsync/mainnet/"
```
Comment (CoderZhi) [inline tools/timemachine/internal/cmd/download.go:33]: support both mainnet and testnet. you can leave it to next pr

File: state/factory/statedb.go [hunk 38]
```diff
@@ -352,6 +361,9 @@ func (sdb *stateDB) PutBlock(ctx context.Context, blk *block.Block) error {
 		)
 	}
 
+	if blk.Height() == sdb.stopHeight {
```
Comment (Liuhaai) [inline state/factory/statedb.go:364]: It's strange to add `stopHeight` inside `stateDB` component. The purpose could be achieved by refactoring the `PutBlock`

File: chainservice/builder.go [hunk 39]
```diff
@@ -127,16 +128,16 @@ func (builder *Builder) createInstance() {
 	}
 }
 
-func (builder *Builder) buildFactory(forTest bool) error {
-	factory, err := builder.createFactory(forTest)
+func (builder *Builder) buildFactory(forTest bool, opts ...factory.StateDBOption) error {
```
Comment (Liuhaai) [inline chainservice/builder.go:131]: mixing option with boolean is misleading

File: tools/timemachine/miniserver/miniServer.go [hunk 40]
```diff
@@ -0,0 +1,132 @@
...
+}
+
+// NewMiniServer creates instace and runs chainservice
+func NewMiniServer(cfg config.Config, opts ...Option) (*MiniServer, error) {
+	svr := &MiniServer{}
+	for _, opt := range opts {
+		opt(svr)
+	}
+
+	builder := chainservice.NewBuilder(cfg)
```
Comment (Liuhaai) [inline tools/timemachine/miniserver/miniServer.go:57]: IMO, `blockDAO` with its sub-components could serve this tool. No need to setup an chain for it.

**Comment by Liuhaai** [comment]
The `Description` of the pr is incomplete

File: state/factory/factory.go [hunk 41]
```diff
@@ -471,6 +480,9 @@ func (sf *factory) PutBlock(ctx context.Context, blk *block.Block) error {
 		)
 	}
 
+	if blk.Height() == sf.stopHeight {
+		return fmt.Errorf("factory catches up to stop height %d", sf.stopHeight)
```
Comment (dustinxie) [inline state/factory/factory.go:484]: it's more accurate to say "chain sync halted right before committing block %d", sf.stopHeight

File: tools/timemachine/internal/cmd/download.go (hunk 37 above)
Comment (huof6829) [reply tools/timemachine/internal/cmd/download.go]: fixed

File: tools/timemachine/internal/cmd/download.go (hunk 36 above)
Comment (huof6829) [reply tools/timemachine/internal/cmd/download.go]: fixed

File: tools/timemachine/internal/cmd/get.go (hunk 35 above)
Comment (huof6829) [reply tools/timemachine/internal/cmd/get.go]: fixed

File: tools/timemachine/internal/cmd/root.go (hunk 34 above)
Comment (huof6829) [reply tools/timemachine/internal/cmd/root.go]: fixed

File: chainservice/builder.go (hunk 33 above)
Comment (huof6829) [reply chainservice/builder.go]: I use `SetStopHeight` to keep consistent with other function like `SetActionPool`

File: chainservice/builder.go (hunk 39 above)
Comment (huof6829) [reply chainservice/builder.go]: I use `SetStopHeight`

File: state/factory/factory.go (hunk 41 above)
Comment (huof6829) [reply state/factory/factory.go]: fixed.

**Comment by huof6829** [comment]
截屏2023-02-09 20 40 11

**Comment by huof6829** [comment]
Glad to receive your reply.Thanks very much！Best  regards~

**Pull request closed without merging**